### ISSUE TITLE
UMAP metrics?

### ISSUE THREAD

**Issue by Fil** [saehm/DruidJS#25]
I was trying to use the jaccard distance as a metric on UMAP, but [it seems that it knows](https://github.com/saehm/DruidJS/blob/master/dimred/UMAP.js#L270) only euclidian_squared and "precomputed". Is it something you'd be willing to add?

~~~js
function jaccard(a, b) {
  let c = 0, n = 0;
  for (let i = 0, l = a.length; i < l; ++i) {
    c += a[i] && b[i];
    n += a[i] || b[i]
  }
  return 1 - c / n;
}
~~~

**Comment by saehm** [comment]
If you compute the distance matrix by yourself with the jaccard metric then you can use it, but it does not work right, because there is the gradient from the (reduced?) eucliden distance hard coded. To work properly we would also need a function computing the gradient from the metric. I hope i did not get something wrong.

**Comment by saehm** [comment]
Oh... Sorry, i think i have misunderstood your question. The line you are pointing out is wrong, it is the distance of the embedded points during the optimization.
I had to made [this](https://github.com/saehm/DruidJS/blob/068c443b9cff25a3591c6182edc0b19fac4c6ac8/dimred/UMAP.js#L149) changes, to use different metrics. It should work now with version 0.3.16. I translated also some metrics from the original UMAP implementation (jaccard, sokal_michener, hamming, and yule).
